xen/arch/arm/tee/ffa_partinfo.c

```diff
@@ -78,79 +81,95 @@ int32_t ffa_handle_partition_info_get(uint32_t w1, uint32_t w2, uint32_t w3,
-    {
-        if ( ctx->guest_vers == FFA_VERSION_1_0 )
-            *fpi_size = sizeof(struct ffa_partition_info_1_0);
-        else
-            *fpi_size = sizeof(struct ffa_partition_info_1_1);
-
-        *count = 0;
-        ret = FFA_RET_OK;
-        goto out;
-    }
-
-    if ( !ctx->page_count || !ctx->rx_is_free )
-        goto out;
-    spin_lock(&ffa_rx_buffer_lock);
-    ret = ffa_partition_info_get(w1, w2, w3, w4, w5, count, fpi_size);
-    if ( ret )
-        goto out_rx_buf_unlock;
     /*
-     * ffa_partition_info_get() succeeded so we now own the RX buffer we
-     * share with the SPMC. We must give it back using ffa_rx_release()
-     * once we've copied the content.
+     * If the guest is v1.0, he does not get back the entry size so we must
+     * use the v1.0 structure size in the destination buffer.
+     * Otherwise use the size of the highest version we support, here 1.1.
      */
-
     if ( ctx->guest_vers == FFA_VERSION_1_0 )
+        dst_size = sizeof(struct ffa_partition_info_1_0);
+    else
+        dst_size = sizeof(struct ffa_partition_info_1_1);
+
+    dst_buf = ctx->rx;
+
+    if ( ffa_fw_support_fid(FFA_PARTITION_INFO_GET) )
     {
-        size_t n;
-        struct ffa_partition_info_1_1 *src = ffa_rx;
-        struct ffa_partition_info_1_0 *dst = ctx->rx;
+        ret = FFA_RET_DENIED;
 
-        if ( ctx->page_count * FFA_PAGE_SIZE < *count * sizeof(*dst) )
-        {
-            ret = FFA_RET_NO_MEMORY;
-            goto out_rx_release;
-        }
+        if ( !ffa_rx )
+            goto out;
+
+        if ( !ctx->page_count || !ctx->rx_is_free )
+            goto out;
+
+        spin_lock(&ffa_rx_buffer_lock);
 
-        for ( n = 0; n < *count; n++ )
+        ret = ffa_partition_info_get(w1, w2, w3, w4, w5, &num_ffa_sp,
+                                     &src_size);
+
+        if ( ret )
+            goto out_rx_buf_unlock;
+
+        /*
+         * ffa_partition_info_get() succeeded so we now own the RX buffer we
+         * share with the SPMC. We must give it back using ffa_rx_release()
+         * once we've copied the content.
+         */
+
+        /* we cannot have a size lower than 1.0 structure */
+        if ( src_size < sizeof(struct ffa_partition_info_1_0) )
         {
-            dst[n].id = src[n].id;
-            dst[n].execution_context = src[n].execution_context;
-            dst[n].partition_properties = src[n].partition_properties;
+            ret = FFA_RET_NOT_SUPPORTED;
+            goto out_rx_release;
         }
     }
     else
     {
-        size_t sz = *count * *fpi_size;
+        ret = FFA_RET_OK;
+        *count = 0;
+        *size = dst_size;
+        goto out;
+    }
+
+    if ( ctx->page_count * FFA_PAGE_SIZE < num_ffa_sp * dst_size )
+    {
+        ret = FFA_RET_NO_MEMORY;
+        goto out_rx_release;
+    }
 
-        if ( ctx->page_count * FFA_PAGE_SIZE < sz )
+    if ( num_ffa_sp > 0 )
+    {
+        uint32_t n;
+        void *src_buf = ffa_rx;
```

Comment:
Since we're doing arithmetic on this pointer, should it be a `uint8_t *`?

Reply:
same as upper, arithmetic on void is supported and defined